File: solutions/0003-longest-substring-without-repeating-characters/solution.java
// import java.util.Scanner;

// class Solution {
//     public static int lengthOfLongestSubstring(String s) {
//         char[]ch = s.toCharArray();
//         ArrayList<Character> arr = new ArrayList<>();
//         int maxsum = 0;
//         int index = 0;
//         for(int i=index;i<s.length();i++){
//             if(!arr.contains(ch[i])){
//                 arr.add(ch[i]);
//             }
//             else{
//                 if(arr.size()>maxsum){
//                 maxsum = arr.size();
//                 }
//                 index++;
//                 arr.clear();
//             }
//         }
//         return maxsum;
//     }
//     public static void main(String[]args){
//         Scanner sc = new Scanner(System.in);
//         String s = sc.next();
//         lengthOfLongestSubstring(s);
//     }
// }

import java.util.ArrayList;
import java.util.Scanner;

class Solution {
    public static int lengthOfLongestSubstring(String s) {
        char[] ch = s.toCharArray();
        ArrayList<Character> arr = new ArrayList<>();
        int maxsum = 0;
        int index = 0;
        
        // Iterate over the string
        for (int i = 0; i < s.length(); i++) {
            if (!arr.contains(ch[i])) {
                // Add the character if it's not already in the substring
                arr.add(ch[i]);
            } else {
                if (arr.size() > maxsum) {
                    maxsum = arr.size();
                }

                arr.clear();
                index++;  // Move to the next starting character
                i = index - 1;  // Reset 
            }
        }
        
        // Final check after the loop
        if (arr.size() > maxsum) {
            maxsum = arr.size();
        }
        
        return maxsum;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String s = sc.next();  // Input string
        int result = lengthOfLongestSubstring(s);  // Get the result
        System.out.println(result);  // Print the result
    }
}

File: solutions/0003-longest-substring-without-repeating-characters/solution.py
class Solution:
    def lengthOfLongestSubstring(
        self, s: str, currmaxstring: str = "",
        maxstring: str = "", index: int = 0
    ) -> int:

        if len(s) <= 1:
            return len(s)

        if index >= len(s):
            return len(maxstring)

        for i in range(index, len(s)):

            if s[i] in currmaxstring:

                if len(currmaxstring) > len(maxstring):
                    maxstring = currmaxstring

                index += 1
                currmaxstring = ""

                return self.lengthOfLongestSubstring(
                    s, currmaxstring, maxstring, index
                )

            else:
                currmaxstring += s[i]

        return max(len(currmaxstring), len(maxstring))
